src/pkg/packager/composer/list.go

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package composer contains functions for composing components within Zarf packages.
+package composer
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/defenseunicorns/zarf/src/pkg/layout"
+	"github.com/defenseunicorns/zarf/src/pkg/oci"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
+	"github.com/defenseunicorns/zarf/src/types"
+)
+
+// Node is a node in the import chain
+type Node struct {
+	types.ZarfComponent
+
+	prev *Node
+	next *Node
+}
+
+// ImportChain is a doubly linked list of components
+type ImportChain struct {
+	head *Node
+	tail *Node
+}
+
+func (ic *ImportChain) append(c types.ZarfComponent) {
+	node := &Node{ZarfComponent: c, prev: nil, next: nil}
+	if ic.head == nil {
+		ic.head = node
+		ic.tail = node
+	} else {
+		p := ic.head
+		for p.next != nil {
+			p = p.next
+		}
+		node.prev = p
+
+		p.next = node
+		ic.tail = node
+	}
+}
+
+// NewImportChain creates a new import chain from a component
+func NewImportChain(head types.ZarfComponent, arch string) (*ImportChain, error) {
+	ic := &ImportChain{}
+
+	ic.append(head)
+
+	history := []string{}
+
+	node := ic.head
+	for node != nil {
+		isLocal := node.Import.Path != "" && node.Import.URL == ""
+		isRemote := node.Import.Path == "" && node.Import.URL != ""
+
+		if !isLocal && !isRemote {
+			// EOL
+			return ic, nil
+		}
+
+		if node.prev != nil && node.prev.Import.URL != "" {
+			return ic, fmt.Errorf("detected malformed import chain, cannot import remote components from remote components")
+		}
+
+		var pkg types.ZarfPackage
+		name := node.Name
+
+		if isLocal {
+			history = append(history, node.Import.Path)
+			paths := append(history, layout.ZarfYAML)
+			if err := utils.ReadYaml(filepath.Join(paths...), &pkg); err != nil {
+				return ic, err
+			}
+		} else if isRemote {
+			remote, err := oci.NewOrasRemote(node.Import.URL)
+			if err != nil {
+				return ic, err
+			}
+			pkg, err = remote.FetchZarfYAML()
+			if err != nil {
+				return ic, err
+			}
+		}
+
+		if node.Import.ComponentName != "" {
+			name = node.Import.ComponentName
+		}
+
+		found := helpers.Find(pkg.Components, func(c types.ZarfComponent) bool {
+			return c.Name == name
+		})
+
+		if found.Name == "" {
+			if isLocal {
+				return ic, fmt.Errorf("component %q not found in %q", name, filepath.Join(history...))
+			} else if isRemote {
+				return ic, fmt.Errorf("component %q not found in %q", name, node.Import.URL)
+			}
+		}
+
+		if node.Only.Cluster.Architecture != "" {
+			arch = node.Only.Cluster.Architecture
+		}
+
+		if arch != "" && found.Only.Cluster.Architecture != "" && found.Only.Cluster.Architecture != arch {
```

Comment:
This isn't quite how this should work - the imported package may have multiple components of the same name with different archs so this logic should be in the found function instead.  See this for an example:

https://github.com/Racer159/zarf-package-k3d/blob/master/install/linux/zarf.yaml

Reply:
uhh https://github.com/defenseunicorns/zarf/blob/6a04e7783d1308cb22eda22a1ebca4dcbdcdf2e0/src/internal/packager/validate/validate.go#L56